src/crypto/random.cpp

```diff
@@ -0,0 +1,189 @@
+// 
+// 2. Modification of the source form for your own personal use. 
+// 
+// As long as the following conditions are met: 
+// 
+// 3. You must not distribute modified copies of the work to third parties. This includes 
+//    posting the work online, or hosting copies of the modified work for download. 
+// 
+// 4. Any derivative version of this work is also covered by this license, including point 8. 
+// 
+// 5. Neither the name of the copyright holders nor the names of the authors may be 
+//    used to endorse or promote products derived from this software without specific 
+//    prior written permission. 
+// 
+// 6. You agree that this licence is governed by and shall be construed in accordance 
+//    with the laws of England and Wales. 
+// 
+// 7. You agree to submit all disputes arising out of or in connection with this licence 
+//    to the exclusive jurisdiction of the Courts of England and Wales. 
+// 
+// Authors and copyright holders agree that: 
+// 
+// 8. This licence expires and the work covered by it is released into the 
+//    public domain on 1st of February 2019 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL 
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF 
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+
+#include <string.h>
+#include <iostream>
+
+#include "random.hpp"
+#include "keccak.h"
+
+#if defined(_WIN32)
+
+#include <stdio.h>
+#include <windows.h>
+#include <wincrypt.h>
+
+#else
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#endif
+
+extern "C" void* memwipe(void *src, size_t n);
+
+struct prng_handle
+{
+#if defined(_WIN32)
+	HCRYPTPROV prov;
+#else
+	int fd;
+#endif
+};
+
+prng::~prng()
+{
+	if(hnd == nullptr)
+		return;
+
+#if defined(_WIN32)
+	if(!CryptReleaseContext(hnd->prov, 0))
+	{
+		std::cerr << "CryptReleaseContext" << std::endl;
+		std::abort();
+	}
+#else
+	if(close(hnd->fd) < 0)
+	{
+		std::cerr << "Exit Failure :: close /dev/urandom " << std::endl; 
+		std::abort();
+	}
+#endif
+	delete hnd;
+}
+
+void prng::start()
+{
+	hnd = new prng_handle;
+#if defined(_WIN32)
+	if(!CryptAcquireContext(&hnd->prov, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
```

Comment:
Why not using standard c++ and use https://en.cppreference.com/w/cpp/numeric/random/random_device/random_device
This will avoid that we need to fiddle around with platform specific methods to access the random number generator.

Reply:
This is security critical. We won't use anything not crypto-grade.